test/test_core.h

```diff
@@ -28,16 +28,80 @@
 #include "picotest/picotest.h"
 #include "testhelper.h"
 #include "tiny_dnn/tiny_dnn.h"
-
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+#include "third_party/CLCudaAPI/clpp11.h"
+#endif //defined(USE_OPENCL) || defined(USE_CUDA)
 using namespace tiny_dnn;
 
 namespace tiny_dnn {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+device_t device_type(size_t &platform, size_t &device) {
+    //check which platforms are available
+    auto platforms = CLCudaAPI::GetAllPlatforms();
+
+    // if no platforms - return -1
+    if (platforms.size() == 0)
+        return device_t::NONE;
+
+    for (auto p = platforms.begin(); p != platforms.end(); ++p)
+        for (size_t d = 0; d < p->NumDevices(); ++d) {
+            auto dev = CLCudaAPI::Device(*p, d);
+            if (dev.Type()=="GPU") {
+                platform = p - platforms.begin();
+                device = d;
+                return device_t::GPU;
+            }
+        }
+
+    for (auto p = platforms.begin(); p != platforms.end(); ++p)
+        for (size_t d = 0; d < p->NumDevices(); ++d) {
+            auto dev = CLCudaAPI::Device(*p, d);
+            if (dev.Type()=="CPU") {
+                platform = p - platforms.begin();
+                device = d;
+                return device_t::CPU;
+            }
+        }
+
```

Comment:
Another option is:
```cpp
	auto devices_order = {"FPGA" /* future */, "GPU", "CPU"}
	auto devices_t_order = {device_t::FPGA /* future */, device_t::GPU, device_t::CPU}
	for (size_t d_type = 0; d_type < devices_order.size(); ++d_type)
		for (auto p = platforms.begin(); p != platforms.end(); ++p)
			for (size_t d = 0; d < p->NumDevices(); ++d) {
				auto dev = CLCudaAPI::Device(*p, d);
				if (dev.Type() == devices_order[d_type]) {
					platform = p - platforms.begin();
					device = d;
					return devices_t_order[d_type];
				}
			}
	return device_t::NONE;
```
Also we can give different priority to devices with `define`s.